%% Generated by Sphinx.
\def\sphinxdocclass{report}
\documentclass[letterpaper,12pt,english]{sphinxmanual}
\ifdefined\pdfpxdimen
   \let\sphinxpxdimen\pdfpxdimen\else\newdimen\sphinxpxdimen
\fi \sphinxpxdimen=.75bp\relax

\usepackage[utf8]{inputenc}
\ifdefined\DeclareUnicodeCharacter
 \ifdefined\DeclareUnicodeCharacterAsOptional
  \DeclareUnicodeCharacter{"00A0}{\nobreakspace}
  \DeclareUnicodeCharacter{"2500}{\sphinxunichar{2500}}
  \DeclareUnicodeCharacter{"2502}{\sphinxunichar{2502}}
  \DeclareUnicodeCharacter{"2514}{\sphinxunichar{2514}}
  \DeclareUnicodeCharacter{"251C}{\sphinxunichar{251C}}
  \DeclareUnicodeCharacter{"2572}{\textbackslash}
 \else
  \DeclareUnicodeCharacter{00A0}{\nobreakspace}
  \DeclareUnicodeCharacter{2500}{\sphinxunichar{2500}}
  \DeclareUnicodeCharacter{2502}{\sphinxunichar{2502}}
  \DeclareUnicodeCharacter{2514}{\sphinxunichar{2514}}
  \DeclareUnicodeCharacter{251C}{\sphinxunichar{251C}}
  \DeclareUnicodeCharacter{2572}{\textbackslash}
 \fi
\fi
\usepackage{cmap}
\usepackage[T1]{fontenc}
\usepackage{amsmath,amssymb,amstext}
\usepackage{babel}
\usepackage{times}
\usepackage[Bjarne]{fncychap}
\usepackage[dontkeepoldnames]{sphinx}

\usepackage{geometry}

% Include hyperref last.
\usepackage{hyperref}
% Fix anchor placement for figures with captions.
\usepackage{hypcap}% it must be loaded after hyperref.
% Set up styles of URL: it should be placed after hyperref.
\urlstyle{same}
\addto\captionsenglish{\renewcommand{\contentsname}{Contents:}}

\addto\captionsenglish{\renewcommand{\figurename}{Fig.}}
\addto\captionsenglish{\renewcommand{\tablename}{Table}}
\addto\captionsenglish{\renewcommand{\literalblockname}{Listing}}

\addto\captionsenglish{\renewcommand{\literalblockcontinuedname}{continued from previous page}}
\addto\captionsenglish{\renewcommand{\literalblockcontinuesname}{continues on next page}}

\addto\extrasenglish{\def\pageautorefname{page}}

\setcounter{tocdepth}{1}



\title{quartetsampling Documentation}
\date{Aug 05, 2019}
\release{1.3.0}
\author{James B. Pease \\ Stephen A. Smith \\ Cody E. Hinchliff \\Joseph W. Brown \\ Joseph F. Walker}
\newcommand{\sphinxlogo}{\sphinxincludegraphics{logo.png}\par}
\renewcommand{\releasename}{Release}
\makeindex

\begin{document}

\maketitle
\sphinxtableofcontents
\phantomsection\label{\detokenize{index::doc}}



\chapter{Getting Started}
\label{\detokenize{intro:manual-for-quartetsampling}}\label{\detokenize{intro::doc}}\label{\detokenize{intro:intro}}\label{\detokenize{intro:getting-started}}

\section{What is Quartet Sampling?}
\label{\detokenize{intro:what-is-quartet-sampling}}
Quartet Sampling (QS) is a method for quantifying branch support values for a phylogenetic tree.  The software requires an input tree topology (branch lengths not required or used) and a molecular alignment in phylip format.  Optionally, a RAxML-formatted partition file may be used for either a concatenated alignment in partitioned mode or to establish gene barriers for using individual gene trees.  The Quartet Sampling method takes a phylogenetic topology and considers each internal (i.e., non-terminal) branch.  Each internal branch has four branches connected to it that lead to one or more taxa.  For each replicate, Quartet Sampling randomly selects one taxon from each of the four groups and evaluates the likelihoods of the three possible topological arrangements of the four taxa.  These replicates then form the basis of branch scores for the QS method.  Full details about the scientific background and equations behind QS can be found in the Quartet Sampling paper in the American Journal of Botany at \textless{}\sphinxurl{http://dx.doi.org/10.1002/ajb2.1016}\textgreater{}.


\section{How do I cite this program?}
\label{\detokenize{intro:how-do-i-cite-this-program}}
James B Pease, Joseph W Brown, Joseph F Walker, Cody E Hinchliff, Stephen A Smith. 2018. Quartet Sampling distinguishes lack of support from conflicting support in the green plant tree of life. American Journal of Botany. 105(3): 385\textendash{}403. doi:10.1002/ajb2.1016

Please also include the URL \textless{}\sphinxurl{https://www.github.com/fephyfofum/quartetsampling}\textgreater{} in your methods section where the program is referenced.

\sphinxstylestrong{*Please also be sure to cite RAxML-ng, RAxML, PAUP, or IQ-TREE*} depending on which engine you use (the default is RAxML-ng).


\section{Installation}
\label{\detokenize{intro:installation}}
No installation is required, quartetsampling scripts should work as long as Python and RAxML are installed.  The repository can be cloned or downloaded as a .zip file from GitHub.


\subsection{Requirements}
\label{\detokenize{intro:requirements}}\begin{itemize}
\item {} 
Python 3.x (2.7 may also work, but \sphinxstylestrong{3.x strongly recommended} for optimal parallelization) \sphinxurl{https://www.python.org/downloads/}

\item {} 
RAxML-ng 0.9.0 (beta versions not guaranteed) \sphinxurl{https://github.com/amkozlov/raxml-ng}

\end{itemize}

\sphinxstyleemphasis{Alternative likelihood engines available}
* RAxML 8.1+ (8.0.x and 7.x will not work) \sphinxurl{https://sco.h-its.org/exelixis/web/software/raxml/index.html}
* PAUP  \sphinxurl{http://people.sc.fsu.edu/~dswofford/paup\_test/}
* IQ-TREE 1.6.x (earlier versions may not work) \sphinxurl{http://www.iqtree.org}

\sphinxstyleemphasis{Optional, but recommended:*}
\begin{itemize}
\item {} 
Numpy \sphinxurl{http://www.numpy.org} (required only for \sphinxcode{calc\_qs\_stats.py})

\item {} 
Scipy \sphinxurl{https://www.scipy.org} (required only if \sphinxcode{-{-}calc-qdstats} is activated)

\item {} 
Figtree \sphinxurl{http://tree.bio.ed.ac.uk/software/figtree/} (view FigTree output)

\end{itemize}


\section{Preparing your data}
\label{\detokenize{intro:preparing-your-data}}

\subsection{Phylogeny}
\label{\detokenize{intro:phylogeny}}
A phylogenetic tree in Newick (parenthetical) format should be used.  Branch lengths are optional and will be ignored by the program.  Removal of support scores in square brackets is recommended.

\begin{sphinxadmonition}{note}{Note:}
Internal branch labels will be replaced in output trees.
\end{sphinxadmonition}


\subsection{Sequence Alignment}
\label{\detokenize{intro:sequence-alignment}}
An alignment in Relaxed Phylip format (such as used for RAxML) is required.  The alignment can be DNA nucleotides or amino acids (use \sphinxcode{-{-}amino-acid} in that case). If you have an alignment in FASTA format, we also include a script called \sphinxcode{utils/fasta2phy.py} that will convert FASTA alignments to Relaxed Phylip.

\begin{sphinxadmonition}{important}{Important:}
Labels for the alignment sequences must match the labels on tree terminal branches exactly. All tips in the phylogeny must have a sequence represented in the alignment.  Sequences appearing in the alignment, but not in the tree are allowed, and will be ignored.
\end{sphinxadmonition}


\subsection{Partitions File (Optional)}
\label{\detokenize{intro:partitions-file-optional}}
A partitions file in the style of RAxML may also be used either for a partitioned analysis of likelihood or separate gene tree evaluation.  DNA partitions should use the \sphinxcode{DNA} prefix, and proteins should use the \sphinxcode{WAG} prefix instead of \sphinxcode{DNA} in the example below.  Note that partition names are arbitrary in this case and that number ranges are inclusive.
\begin{description}
\item[{Example file::}] \leavevmode
DNA, p1=1-30
DNA, p2=31-60
DNA, p3=61-90
…

\end{description}

\begin{sphinxadmonition}{important}{Important:}
If your alignment is sparse and you use partitioned most, such that you will frequently have partitions with no sequence data for randomly selected quartets of taxa, you may wish to invoke the \sphinxcode{-{-}ignore-error} option in \sphinxcode{quartet\_sampling.py} to ignore the RAxML errors that will result from these empty partitions.
\end{sphinxadmonition}


\section{Basic usage}
\label{\detokenize{intro:basic-usage}}\begin{description}
\item[{::}] \leavevmode
python3 quartet\_sampling.py \textendash{}tree TREE.nwk \textendash{}align ALIGNMENT.phy \textendash{}reps 100 \textendash{}threads 4 \textendash{}lnlike 2

\end{description}


\subsection{Required Paramters}
\label{\detokenize{intro:required-paramters}}
\sphinxcode{-{-}tree}: File containing a single Newick-formatted phylogeny.

\sphinxcode{-{-}align} Phylip-formatted alignment containing one sequence for each tip in the phylogeny provided.

\sphinxcode{-{-}reps} Number of replicates per branch to perform.

\sphinxcode{-{-}threads} Number of threads for Python to use in parallel (does not pass through to RAxML, this is for Python multiprocessing of per-branch replicates in parallel)


\subsection{Recommended Parameters}
\label{\detokenize{intro:recommended-parameters}}
\sphinxcode{-{-}lnlike}: log-likelihood threshold cutoff, determines the minimum difference by which the best likelihood tree must exceed the second-best likelihood tree when comparing the three possible topologies for a given quartet replicate.

\begin{sphinxadmonition}{note}{Note:}
If \sphinxcode{-{-}lnlike} is omitted, this will invoke an alternative mode where a tree is simply inferred from the sequence data by RAxML (or PAUP*) and likelihoods are not evaluated.  This will result in Quartet Informativeness (QI) scores of ‘NA’ for all branches.
\end{sphinxadmonition}


\chapter{Program Parameter Descriptions}
\label{\detokenize{prog_desc:program-parameter-descriptions}}\label{\detokenize{prog_desc::doc}}

\section{quartet\_sampling}
\label{\detokenize{prog_desc:quartet-sampling}}

\subsection{Description}
\label{\detokenize{prog_desc:description}}
quartet\_samping.py: Quartet Sampling method for
phylogenetic branch support evaluation

\textless{}\sphinxurl{http://www.github.com/FePhyFoFum/quartetsampling}\textgreater{}


\subsection{Parameters}
\label{\detokenize{prog_desc:parameters}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:h-help}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}align/-{-}alignment} (required)}
\label{\detokenize{prog_desc:align-alignment-required}}
\sphinxstylestrong{Description:} Alignment file in “relaxed phylip” format, as used by RAxML.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}reps/-{-}number-of-reps} (required)}
\label{\detokenize{prog_desc:reps-number-of-reps-required}}
\sphinxstylestrong{Description:} The number of replicate quartet topology searches to be performed at each node.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 100


\subsubsection{\sphinxstyleliteralintitle{-{-}threads/-{-}number-of-threads} (required)}
\label{\detokenize{prog_desc:threads-number-of-threads-required}}
\sphinxstylestrong{Description:} The number of parallel threads to be used by Python for quartet topology searches.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 1


\subsubsection{\sphinxstyleliteralintitle{-{-}tree} (required)}
\label{\detokenize{prog_desc:tree-required}}
\sphinxstylestrong{Description:} The input tree in Newick (parenthetical) format.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}calc-qdstats}}
\label{\detokenize{prog_desc:calc-qdstats}}
\sphinxstylestrong{Description:} EXPERIMENTAL: Calculates Chi-square test for QD tree frequencies. Use only  if Scipy is available. Will increase running time.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}clade}}
\label{\detokenize{prog_desc:clade}}
\sphinxstylestrong{Description:} Conduct analysis on specific clade identified by CSV taxon list

\sphinxstylestrong{Type:} string; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}data-type}}
\label{\detokenize{prog_desc:data-type}}
\sphinxstylestrong{Description:} (nuc)leotide, (amino) acid, or (cat)egorical data

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} {[}‘nuc’{]}

\sphinxstylestrong{Choices:} (‘nuc’, ‘amino’, ‘cat’)


\subsubsection{\sphinxstyleliteralintitle{-{-}engine}}
\label{\detokenize{prog_desc:engine}}
\sphinxstylestrong{Description:} Name of the program to use to infer trees or evaluate tree model likelihoods.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} (‘raxml-ng’,)

\sphinxstylestrong{Choices:} (‘raxml-ng’, ‘raxml’, ‘paup’, ‘iqtree’)


\subsubsection{\sphinxstyleliteralintitle{-{-}engine-exec}}
\label{\detokenize{prog_desc:engine-exec}}
\sphinxstylestrong{Description:} Full file path of the tree inference or likelihood evaluation engine.

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}engine-model}}
\label{\detokenize{prog_desc:engine-model}}
\sphinxstylestrong{Description:} Advanced: specify a custom model name for the tree engine

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}genetrees}}
\label{\detokenize{prog_desc:genetrees}}
\sphinxstylestrong{Description:} Use partitions file (RAxML format) to divide the alignment into separate gene tree regions. Gene alignments will be sampled random for the quartet topology searches.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}ignore-errors}}
\label{\detokenize{prog_desc:ignore-errors}}
\sphinxstylestrong{Description:} Ignore RAxML and PAUP erroneous runs

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}lnlike/-{-}lnlike-thresh}}
\label{\detokenize{prog_desc:lnlike-lnlike-thresh}}
\sphinxstylestrong{Description:} The lnlike threshhold that is the minimum value by which the log-likelihood value of the best-likelihood tree must be higher than the second-best-likelihood tree for the replicate to register as the best-likelihood topology rather than ‘uncertain’. If set to zero, this turns off likelihood evaluation mode and invokes tree inference mode where a tree is simply inferred from the alignment without considering likelihood (QI values are N/A in this case).

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} 2.0


\subsubsection{\sphinxstyleliteralintitle{-{-}low-mem}}
\label{\detokenize{prog_desc:low-mem}}
\sphinxstylestrong{Description:} Do not store large alignment in memory for whole-alignment (non-genetree) mode

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}max-random-sample-proportion}}
\label{\detokenize{prog_desc:max-random-sample-proportion}}
\sphinxstylestrong{Description:} The proportion of possible replicates explored unsuccessfully by the random generation procedure before it gives up. Because this generates random replicates, it takes progressively longer as it proceeds. To avoid long runtimes, the recommended range is \textless{} 0.5 (which is the default).

\sphinxstylestrong{Type:} float; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}min-overlap}}
\label{\detokenize{prog_desc:min-overlap}}
\sphinxstylestrong{Description:} The minimum sites required to be sampled for all taxa in a given quartet.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}partitions}}
\label{\detokenize{prog_desc:partitions}}
\sphinxstylestrong{Description:} Partitions file in RAxML format. If omitted then the entire alignment will be treated as one partition for all quartet replicate topology searches.

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}result-prefix}}
\label{\detokenize{prog_desc:result-prefix}}
\sphinxstylestrong{Description:} A prefix to put on the result files.

\sphinxstylestrong{Type:} string; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}results-dir}}
\label{\detokenize{prog_desc:results-dir}}
\sphinxstylestrong{Description:} A directory to which output files will be saved. If not supplied, the current working directory will be used. (default is current folder).

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}retain-temp}}
\label{\detokenize{prog_desc:retain-temp}}
\sphinxstylestrong{Description:} Do not remove temporary files

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}start-node-number}}
\label{\detokenize{prog_desc:start-node-number}}
\sphinxstylestrong{Description:} An integer denoting the node to which to start from. Nodes will be read from topologically identical (and isomorphic!) input trees in deterministic order, so this argument may be  used to restart at an intermediate position (in case the previous run was canceled before completion, for example).

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}stop-node-number}}
\label{\detokenize{prog_desc:stop-node-number}}
\sphinxstylestrong{Description:} An integer denoting the node at which to stop. Will include nodes with indices \textless{}= the stop node number. This argument may be used to limit the length of a given run in case only a certain part of the tree is of interest. Nodes will be read from topologically identical (and isomorphic!) input trees in deterministic order.

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}temp-dir}}
\label{\detokenize{prog_desc:temp-dir}}
\sphinxstylestrong{Description:} A directory to which temporary files will be saved. If not supplied, ‘QuartetSampling’ will be created in the current working directory. When specifying a custom temporary output the characters ‘QuartetSampling’ must appear in the directory name to prevent accidental file deletion. (default=’./QuartetSampling’

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}verbose}}
\label{\detokenize{prog_desc:verbose}}
\sphinxstylestrong{Description:} Provide more verbose output if specified.

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}verbout}}
\label{\detokenize{prog_desc:verbout}}
\sphinxstylestrong{Description:} Provide output of the frequencies of each topology and QC.

\sphinxstylestrong{Type:} boolean flag


\section{merge\_output}
\label{\detokenize{prog_desc:merge-output}}

\subsection{Description}
\label{\detokenize{prog_desc:id1}}
Combines RESULT.node.scores.csv files from separate
runs for the same phylogeny into a single set of csv and tree outputs.

\sphinxurl{http://www.github.com/FePhyFoFum/quartetsampling}


\subsection{Parameters}
\label{\detokenize{prog_desc:id2}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id3}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}nodedata} (required)}
\label{\detokenize{prog_desc:nodedata-required}}
\sphinxstylestrong{Description:} file containing paths of one or moreRESULT.node.score.csv files

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}out} (required)}
\label{\detokenize{prog_desc:out-required}}
\sphinxstylestrong{Description:} new output files prefix

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}tree} (required)}
\label{\detokenize{prog_desc:id4}}
\sphinxstylestrong{Description:} tree file in Newick format

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}clade}}
\label{\detokenize{prog_desc:id5}}
\sphinxstylestrong{Description:} ==SUPPRESS==

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}startk}}
\label{\detokenize{prog_desc:startk}}
\sphinxstylestrong{Description:} ==SUPPRESS==

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 0


\subsubsection{\sphinxstyleliteralintitle{-{-}stopk}}
\label{\detokenize{prog_desc:stopk}}
\sphinxstylestrong{Description:} ==SUPPRESS==

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}verbose}}
\label{\detokenize{prog_desc:id6}}
\sphinxstylestrong{Description:} None

\sphinxstylestrong{Type:} boolean flag


\section{query\_tree}
\label{\detokenize{prog_desc:query-tree}}

\subsection{Description}
\label{\detokenize{prog_desc:id7}}
Tree query script to find specific nodes numbers in large trees
when using the post-run annotated trees.

\sphinxurl{http://www.github.com/FePhyFoFum/quartetsampling}


\subsection{Parameters}
\label{\detokenize{prog_desc:id8}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id9}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}clade}}
\label{\detokenize{prog_desc:id10}}
\sphinxstylestrong{Description:} ==SUPPRESS==

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}data}}
\label{\detokenize{prog_desc:data}}
\sphinxstylestrong{Description:} CSV output from quartet\_sampling (RESULT.node.score.csv)

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}startk}}
\label{\detokenize{prog_desc:id11}}
\sphinxstylestrong{Description:} ==SUPPRESS==

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 0


\subsubsection{\sphinxstyleliteralintitle{-{-}stopk}}
\label{\detokenize{prog_desc:id12}}
\sphinxstylestrong{Description:} ==SUPPRESS==

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}tree}}
\label{\detokenize{prog_desc:tree}}
\sphinxstylestrong{Description:} input tree in newick format

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}verbose}}
\label{\detokenize{prog_desc:id13}}
\sphinxstylestrong{Description:} verbose screen output

\sphinxstylestrong{Type:} boolean flag


\section{calc\_qstats}
\label{\detokenize{prog_desc:calc-qstats}}

\subsection{Description}
\label{\detokenize{prog_desc:id14}}\begin{description}
\item[{Calculate basic statistics on the}] \leavevmode
RESULTS.node.score.csv output file
from quartet\_sampling

\end{description}


\subsection{Parameters}
\label{\detokenize{prog_desc:id15}}

\subsubsection{\sphinxstyleliteralintitle{-h/-{-}help}}
\label{\detokenize{prog_desc:id16}}
\sphinxstylestrong{Description:} show this help message and exit

\sphinxstylestrong{Type:} boolean flag


\subsubsection{\sphinxstyleliteralintitle{-{-}data} (required)}
\label{\detokenize{prog_desc:data-required}}
\sphinxstylestrong{Description:} RESULT.node.score.csv file output fromquartet\_sampling.py

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}clade}}
\label{\detokenize{prog_desc:id17}}
\sphinxstylestrong{Description:} specify a clade using a comma-separatedlist of 2+ descendant taxa

\sphinxstylestrong{Type:} None; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}out}}
\label{\detokenize{prog_desc:out}}
\sphinxstylestrong{Description:} output file path for statistics

\sphinxstylestrong{Type:} file path; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}startk}}
\label{\detokenize{prog_desc:id18}}
\sphinxstylestrong{Description:} starting branch numerical index

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} 0


\subsubsection{\sphinxstyleliteralintitle{-{-}stopk}}
\label{\detokenize{prog_desc:id19}}
\sphinxstylestrong{Description:} stopping branch numerical index

\sphinxstylestrong{Type:} integer; \sphinxstylestrong{Default:} None


\subsubsection{\sphinxstyleliteralintitle{-{-}verbose}}
\label{\detokenize{prog_desc:id20}}
\sphinxstylestrong{Description:} verbose screen output

\sphinxstylestrong{Type:} boolean flag


\chapter{Outputs}
\label{\detokenize{output::doc}}\label{\detokenize{output:output}}\label{\detokenize{output:outputs}}

\section{RESULT.node.scores.csv}
\label{\detokenize{output:result-node-scores-csv}}
A comma-separated values (CSV) file with
* \sphinxstylestrong{node\_label} The label of the internal branch (QS\#\#\#) or terminal branch (original label)
* \sphinxstylestrong{freq0} The number of concordant replicates over the non-uncertain total
* \sphinxstylestrong{qc} The Quartet Concordance score (internal branches only; measures frequency of concordant over discordant)
* \sphinxstylestrong{qd} The Quartet Differential score (internal branches only; measures skew in the two discordant tree counts)
* \sphinxstylestrong{qi} The Quartet Informativeness score (internal branches only; measures number of replicates that fail likelihood cutoff)
* \sphinxstylestrong{qf} The Quartet Fidelity score (terminal branches only; measures the number of replicates for which this branch produced concordant quartets)
* \sphinxstylestrong{diff} Example likelihood differentinal from last replicate (used for diagnostic purposes)
* \sphinxstylestrong{num\_replicates} Number of replicates actually sampled per branch (may not equal the number specified by \sphinxcode{-N} for internal branches when fewer replicates are possible, and will not equal that number for terminal branches).
* \sphinxstylestrong{notes} Additional notes (enables user to add custom notes to their output after running; e.g., labeling key branches)


\section{RESULT.labeled.tre}
\label{\detokenize{output:result-labeled-tre}}
A Newick tree with each internal branch labeled with their QS\#\# identifier.


\section{RESULT.labeled.tre.freq/qc/qd/qu}
\label{\detokenize{output:result-labeled-tre-freq-qc-qd-qu}}
A Newick tree with the each internal branch labeled with frequency of concordant replicates or QC/QD/QI scores.


\section{RESULT.labeled.tre.figtree}
\label{\detokenize{output:result-labeled-tre-figtree}}
A FigTree format phylogeny \textless{}\sphinxurl{http://tree.bio.ed.ac.uk/software/figtree/}\textgreater{} that contains all QS scores and a “score” field with QC/QD/QI for internal branches.


\chapter{Frequently Asked Questions}
\label{\detokenize{faq::doc}}\label{\detokenize{faq:frequently-asked-questions}}\label{\detokenize{faq:faq}}

\section{What is an appropriate \sphinxstyleliteralintitle{-{-}lnlike} value to use?}
\label{\detokenize{faq:what-is-an-appropriate-lnlike-value-to-use}}
The parameter \sphinxcode{-{-}lnlike} or \sphinxcode{-{-}lnlike-thresh} specifies a log-likelihood differential cutoff.  When the three possible topologies for a given quartet are evaluated using molecular data for that quartet by RAxML, three separate likelihood values are generated, which are then log-transformed.  When the cutoff value is specified this means that you want to ensure that the tree with the best likelihood exceeds the next-most likely tree’s likelihood by at least the value of the cutoff.  This prevents the method from selecting from two or three trees with nearly indistinguishable likelihoods.  These quartet replicates that pass the likelihood cutoff are tabulated separated and calculated as the QI score.

On a practical level, increasing the \sphinxcode{-{-}lnlike} parameter will decrease QI and increase the number of replicates counted as part of QC and QD.  Lowering the \sphinxcode{-{-}lnlike} parameter will increase the QI score, but increase the number of potentially arbitrary counts in QC and QD.

More statistically, the log-likelihood threshold is equivalent to specifying a minimum likelihood ratio (subtracting log-likelihoods is equivalent to dividing raw likelihoods). Let the null hypothesis be zero difference between the tree likelihoods (i.e., any difference is random noise). For each of the two likelihood estimates compared, if error is distributed in a standard normal distribution N(0,1), then 95\% of the distribution lies within \textasciitilde{}2 standard deviations. This means the difference in the two log-likelihoods is chi-square distributed with the critical value for 1 degree of freedom at 3.8414 for a 95\% confidence cutoff.  Therefore, log-likelihood values that differ by 3.84 divided by 2 (for a two-tailed test) equals 3.84/2 = 1.92, which is approximately equal to 2.  Therefore, a \sphinxcode{-{-}lnlike} value of 2 conservatively evaluates that the two likelihoods differ significantly at the alpha =0.05 level.


\section{My partitioned data is not working properly, and is a sparse partitioned set.}
\label{\detokenize{faq:my-partitioned-data-is-not-working-properly-and-is-a-sparse-partitioned-set}}
RAxML and RAxML-ng both require that none of data partitions being evaluated are empty.  This creates a unique challenge for quartet samplingon large-sparse matrices with partitioned data, since with sparse data and partitions there might be many partitions of a given quartet that contain insufficient data.  We are working on a solution, but for now recommend using gene-tree mode or unpartitioned on sparse data.


\chapter{Releases}
\label{\detokenize{releases::doc}}\label{\detokenize{releases:id1}}\label{\detokenize{releases:releases}}

\section{v1.3 - Change to RAxML-ng, modernization of flags}
\label{\detokenize{releases:v1-3-change-to-raxml-ng-modernization-of-flags}}\begin{itemize}
\item {} 
\sphinxstylestrong{The default is now RAxML-ng}, though classic RAxML 8.1+ is still available.  Performance of QS in RAxML-ng on test datasets gives similar results, but is much more efficient.

\item {} 
\sphinxstylestrong{Discontinuation of single-letter flags.} Single-letter flags are no longer functional, but have been retained in the comments of the main \sphinxstyleemphasis{quartet\_sampling.py} script to allow forward updating.  This change brings QS into line with evolving commnuity best practices that find better clarity and reproducibility when word-based flags are used exclusively.

\item {} 
\sphinxstylestrong{\textendash{}engine syntax change:} As we expand support for new tree inference and likelihood evaluation engines, this flag is a more flexible alternative to the previous named ones.  See also notes on \sphinxstyleemphasis{\textendash{}engine-exec} and \sphinxstyleemphasis{\textendash{}engine-model}.

\item {} 
Support for IQ-TREE is also available FOR TESTING PURPOSES ONLY!  We have not fully evaluated QS under the IQ-TREE engine, so use this with caution and be sure to document this in your methods.

\end{itemize}


\section{v1.2.1 - Efficiency update}
\label{\detokenize{releases:v1-2-1-efficiency-update}}\begin{itemize}
\item {} 
RAxML now evaluates all three tree configurations in the same program call.  This reduces RAxML calls and increases efficiency.

\item {} 
Clarification in the manual that RAxML 8.1+ is required for the program to execute correctly.

\end{itemize}


\section{v1.2 - Major Update}
\label{\detokenize{releases:v1-2-major-update}}\begin{itemize}
\item {} 
In order to make all four QS components have a “perfect” score of 1, the Quartet Uncertainty (QU) and Quartet Differential (QD) scores have been inverted. QU is now known Quartet as Informativeness (QI) is the inverse measure (where QI = 1 - QU).  QD is also now inverted in scale, so that 1.0 means no differential in the discordant trees, and 0 means all one discdorant option.

\item {} 
This version is consistent with the release of the updated bioRxiv manuscript.

\end{itemize}


\section{v1.1.1}
\label{\detokenize{releases:v1-1-1}}\begin{itemize}
\item {} 
Changed from -A/\textendash{}amino-acids to -d/\textendash{}data-type with ‘nuc’, ‘amino’, and ‘cat’ for binary/categorical data

\item {} 
Amino acid mode should now work in PAUP as well.

\item {} 
Fixed some issues with temp-dir and results-dir that were causing paths to revert to the root

\item {} 
\sphinxstylestrong{the -e/\textendash{}temp-dir parameter folder name now MUST contain ‘QuartetSampling’ to prevent file deletion}

\end{itemize}


\section{v1.1}
\label{\detokenize{releases:v1-1}}
Contains small fixes to enable efficient documentation and minor changes to argument documentation.


\section{v1.0}
\label{\detokenize{releases:v1-0}}
First public release, concurrent with bioRxiv publication.



\renewcommand{\indexname}{Index}
\printindex
\end{document}